src/rust/bitbox-aes/src/lib.rs

```diff
@@ -0,0 +1,3065 @@
+// Copyright 2025 Shift Crypto AG
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![no_std]
+
+#[cfg_attr(test, macro_use)]
+extern crate alloc;
+
+use alloc::vec::Vec;
+
+use hmac::{digest::FixedOutput, Mac, SimpleHmac};
+use sha2::Sha256;
+
+// AES block size.
+const BLOCK_SIZE: usize = 16;
+
+fn pad_length(buf_len: usize) -> usize {
+    BLOCK_SIZE - buf_len % BLOCK_SIZE
+}
+
+fn pkcs7_pad(buf: &mut Vec<u8>) {
+    let pad_len = pad_length(buf.len());
+    buf.resize(buf.len() + pad_len, pad_len as u8);
+}
+
+fn pkcs7_unpad(buf: &mut Vec<u8>) -> Result<(), ()> {
+    if buf.len() % BLOCK_SIZE != 0 {
+        return Err(());
+    }
+    let pad_len = *buf.last().ok_or(())? as usize;
+    if pad_len == 0 || pad_len > BLOCK_SIZE {
+        return Err(());
+    }
+    // The subtraction can't underflow due to the above conditions.
+    buf.truncate(buf.len() - pad_len);
+    Ok(())
+}
+
+fn encrypt(iv: &[u8; 16], key: &[u8; 32], plain: &[u8]) -> Vec<u8> {
+    // Allocate enough space so copying plain into it and padding it does not create copies that
+    // won't get zeroed.
+    let mut padded = zeroize::Zeroizing::new(Vec::with_capacity(pad_length(plain.len())));
+    padded.extend_from_slice(plain);
```

Comment:
It would be nice to add `debug_assert(padded.as_ref().capacity() == padded_len)` after every place it could reallocate to ensure that it doesn't.

Reply:
`debug_assert` is enabled in "non optimized builds". Do we have that? Our debug builds enable DEBUG, but I don't see where/how it translates to Rust build flags.

Reply:
debug_asserts are disabled in both builds in our case. to get a more similar code size of release and debug builds.

Both release and debug includes debug data (like names of symbols). This is not related to debug asserts.

Reply:
unit tests have debug_asserts enabled for example.

Reply:
Amazing, by adding and running into the assert, I found that I had a bug - the capacity should not be `pad_length(plain.len())` but `plain.len() + pad_length(plain.len())`!

I used `assert!()` though not `debug_assert!()`, since `cargo test` actually did not trigger `debug_assert!()` :thinking: